L1Trigger/TrackFindingTMTT/src/TP.cc

```diff
@@ -0,0 +1,192 @@
+#include "SimTracker/Common/interface/TrackingParticleSelector.h"
+#include "L1Trigger/TrackFindingTMTT/interface/TP.h"
+#include "L1Trigger/TrackFindingTMTT/interface/Stub.h"
+#include "L1Trigger/TrackFindingTMTT/interface/Settings.h"
+#include "L1Trigger/TrackFindingTMTT/interface/Utility.h"
+#include "DataFormats/Math/interface/deltaR.h"
+
+using namespace std;
+
+namespace tmtt {
+
+  //=== Store useful info about this tracking particle
+
+  TP::TP(const TrackingParticlePtr& tpPtr, unsigned int index_in_vTPs, const Settings* settings)
+      : trackingParticlePtr_(tpPtr),
+        index_in_vTPs_(index_in_vTPs),
+        settings_(settings),
+        pdgId_(tpPtr->pdgId()),
+        charge_(tpPtr->charge()),
+        mass_(tpPtr->mass()),
+        pt_(tpPtr->pt()),
+        eta_(tpPtr->eta()),
+        theta_(tpPtr->theta()),
+        tanLambda_(1. / tan(theta_)),
+        phi0_(tpPtr->phi()),
+        vx_(tpPtr->vertex().x()),
+        vy_(tpPtr->vertex().y()),
+        vz_(tpPtr->vertex().z()),
+        d0_(vx_ * sin(phi0_) - vy_ * cos(phi0_)),                      // Copied from CMSSW class TrackBase::d0().
+        z0_(vz_ - (vx_ * cos(phi0_) + vy_ * sin(phi0_)) * tanLambda_)  // Copied from CMSSW class TrackBase::dz().
+  {
+    const vector<SimTrack>& vst = tpPtr->g4Tracks();
+    EncodedEventId eid = vst.at(0).eventId();
+    inTimeBx_ = (eid.bunchCrossing() == 0);  // TP from in-time or out-of-time Bx.
+    physicsCollision_ = (eid.event() == 0);  // TP from physics collision or from pileup.
+
+    this->fillUse();        // Fill use_ flag, indicating if TP is worth keeping.
+    this->fillUseForEff();  // Fill useForEff_ flag, indicating if TP is good for tracking efficiency measurement.
+  }
+
+  //=== Fill truth info with association from tracking particle to stubs.
+
+  void TP::fillTruth(const list<Stub>& vStubs) {
+    for (const Stub& s : vStubs) {
+      for (const TP* tp_i : s.assocTPs()) {
+        if (tp_i->index() == this->index())
+          assocStubs_.push_back(&s);
+      }
+    }
+
+    this->fillUseForAlgEff();  // Fill useForAlgEff_ flag.
+
+    this->calcNumLayers();  // Calculate number of tracker layers this TP has stubs in.
+  }
+
+  //=== Check if this tracking particle is worth keeping.
+  //=== (i.e. If there is the slightest chance of reconstructing it, so as to measure fake rate).
+
+  void TP::fillUse() {
+    constexpr bool useOnlyInTimeParticles = false;
+    constexpr bool useOnlyTPfromPhysicsCollisionFalse = false;
+    // Use looser cuts here those those used for tracking efficiency measurement.
+    // Keep only those TP that have a chance (allowing for finite track resolution) of being reconstructed as L1 tracks. L1 tracks not matching these TP will be defined as fake.
+
+    // Include all possible particle types here, as if some are left out, L1 tracks matching one of missing types will be declared fake.
+    static const vector<int> genPdgIdsAllUnsigned = {11, 13, 211, 321, 2212};
```

Comment:
remaining static variable (this one is probably harmless...)

Reply:
Since "static const" data is thread safe, and following the discussion leading to https://github.com/cms-sw/cmssw/pull/29381#issuecomment-628889782 , we propose to leave this code unchanged.

Reply:
I don't mind the `static const`, but why `vector` and not `array` ?

Reply:
done